### PR TITLE
Prepare logical types in validation

### PR DESCRIPTION
**Pull request by Artimi**
Attempt to fix the main issue in https://github.com/fastavro/fastavro/issues/173.

### CODE REVIEW

File: fastavro/write.py [hunk 1]
```diff
@@ -11,7 +12,7 @@
 Writer = _write.Writer
 json_writer = json_write.json_writer
 schemaless_writer = _write.schemaless_writer
-LOGICAL_WRITERS = _write.LOGICAL_WRITERS
+LOGICAL_WRITERS = logical_writers.LOGICAL_WRITERS
```
Comment (Artimi) [inline fastavro/write.py]: I imported it also here to keep the interface of `write` untouched.

File: tests/test_fastavro.py [hunk 2]
```diff
@@ -1898,3 +1899,77 @@ def test_more_null_union_issues():
...
+        {"item": None},
+        {"item": {"item": "2019-05-06"}}
+    ]
+
+    expected = [
+        {"item": None},
+        {"item": {"item": datetime.date(2019, 5, 6)}}
+    ]
+
+    assert expected == roundtrip(schema, records)
```
Comment (Artimi) [inline tests/test_fastavro.py]: I'm not sure if I placed the tests correctly.

File: fastavro/_validation.pyx [hunk 3]
```diff
@@ -79,7 +80,8 @@ cdef inline bint validate_float(datum, schema=None,
 cdef inline bint validate_fixed(datum, dict schema,
                                 str parent_ns='', bint raise_errors=True):
     return (
-        (isinstance(datum, bytes) and len(datum) == schema['size'])
+        ((isinstance(datum, bytes) or isinstance(datum, bytearray))
+         and len(datum) == schema['size'])
```
Comment (Artimi) [inline fastavro/_validation.pyx]: `test_complex_schema` fails without this.

**Comment by scottbelden** [comment]
@Artimi I got busy this weekend and didn't have a chance to take a look at this. I just did and I think the direction seems good. I think I have a few small comments, but before marking those but before mentioning those, I had a quick question about the intention of this change.

Based on the tests you added, it seems like the intention of this PR is to fix the issue when logical types are used in unions. However, the tests say `customLogicalType` which seems a bit inaccurate. However, I wrote a test that used a custom `IPAddress` type and it did work.

```python
def test_custom_logical_type():
    class IPAddress(object):
        def __init__(self, ipaddr):
            self.ip = ipaddr

    from fastavro.logical_writers import LOGICAL_WRITERS

    def prepare_ipaddress_write(data, schema):
        if isinstance(data, IPAddress):
            return data.ip
        else:
            return data

    LOGICAL_WRITERS["string-testipaddress"] = prepare_ipaddress_write

    schema = {
        "name": "test_date_as_string",
        "type": "string",
        "logicalType": "testipaddress",
    }

    data1 = IPAddress("1.1.1.1")
    binary = serialize(schema, data1)
    data2 = deserialize(schema, binary)
    # assert data2 == data1  # This assert would work if we had a custom logical reader
    assert data2 == data1.ip

    data1 = "1.1.1.1"
    binary = serialize(schema, data1)
    data2 = deserialize(schema, binary)
    # assert data2.ip == data1  # This assert would work if we had a custom logical reader
    assert data2 == data1
```

I just wanted to know if the intention of this PR was to resolve both of these issues (which it seems to do). If so, I think this is valuable and we should move forward on it.

**Comment by Artimi** [comment]
@scottbelden Sorry about the confusion. The name `customLogicalType` is indeed inaccurate and should be something like `dateLogicalType` or something general.
Yes, the intention of the PR is to make logical types work also in unions. I think that your test would pass also on the current master branch because this is the way we already use our custom logical type. The only piece missing is the bug when a logical type is in a union.

**Comment by scottbelden** [comment]
Thanks for the clarification. As mentioned, I will mark up a few comments, but overall I think it looks pretty good!

File: fastavro/_logical_writers.pyx [hunk 4]
```diff
@@ -0,0 +1,234 @@
+# cython: language_level=3str
+
+"""Python code for writing AVRO files"""
```
Comment (scottbelden) [inline fastavro/_logical_writers.pyx]: I would remove this line and the comment lines below it. I don't think this line makes sense in this new file and I don't think the comment lines are accurate for this file either (as we had implemented logical types before the apache repo)

File: fastavro/_logical_writers_py.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
+# cython: auto_cpdef=True
+
+"""Python code for writing AVRO files"""
```
Comment (scottbelden) [inline fastavro/_logical_writers_py.py]: I would remove this line and the comment lines below it. I don't think this line makes sense in this new file and I don't think the comment lines are accurate for this file either (as we had implemented logical types before the apache repo)

File: fastavro/_logical_writers.pyx [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+from libc.time cimport tm, mktime
+from cpython.int cimport PyInt_AS_LONG
+from cpython.tuple cimport PyTuple_GET_ITEM
+
+from fastavro import const
+from ._six import long, mk_bits
+from ._timezone import epoch
+
+CYTHON_MODULE = 1  # Tests check this to confirm whether using the Cython code.
```
Comment (scottbelden) [inline fastavro/_logical_writers.pyx]: This isn't needed. The test only checks this value in the `reader` and `writer`

File: fastavro/write.py (hunk 1 above)
Comment (scottbelden) [reply fastavro/write.py]: Makes sense 👍

File: tests/test_fastavro.py (hunk 2 above)
Comment (scottbelden) [reply tests/test_fastavro.py]: I think this is an okay place for them. They could possibly go under `test_validation` but there's not a strong reason to move them there.

**Comment by Artimi** [comment]
@scottbelden Thank you for the review. I made changes according to your suggestions. I also removed the confusing name of the logical type in tests.

**Pull request merged**